### PR TITLE
fix: alt configs model deployment and training issues

### PR DESCRIPTION
**Pull request by malav-shastri**
*Issue #, if available:*

*Description of changes:*
- Model Deployment and training failed for curatedHub customers last week for models with alt configs. 
- Reverting back on this commit fixed the issue locally so investigated this commit for fixing the issue https://github.com/aws/sagemaker-python-sdk/commit/ba7e6a9b3fa6d3b8f9c104d024acec518605e96d

*Testing done:*
- Notebook testing
- Unit tests

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/jumpstart/factory/model.py [hunk 1]
```diff
@@ -259,6 +259,7 @@ def _add_instance_type_to_kwargs(
         sagemaker_session=kwargs.sagemaker_session,
         model_type=kwargs.model_type,
         config_name=kwargs.config_name,
+        hub_arn=kwargs.hub_arn,
```
Comment (bencrabtree) [inline src/sagemaker/jumpstart/factory/model.py:262]: We're going to be finding these lines forever aren't we....

File: src/sagemaker/jumpstart/types.py [hunk 2]
```diff
@@ -1278,7 +1278,7 @@ def from_json(self, json_obj: Dict[str, Any]) -> None:
             json_obj (Dict[str, Any]): Dictionary representation of spec.
         """
         if self._is_hub_content:
-            json_obj = walk_and_apply_json(json_obj, camel_to_snake)
+            json_obj = walk_and_apply_json(json_obj, camel_to_snake, ["metrics"])
```
Comment (bencrabtree) [inline src/sagemaker/jumpstart/types.py:1281]: nit- you can remove `["metrics"]` since it is the default value

**Review by bencrabtree — APPROVED** [review]

**Review by bencrabtree — APPROVED** [review]

File: src/sagemaker/jumpstart/artifacts/metric_definitions.py [hunk 3]
```diff
@@ -96,16 +96,17 @@ def _retrieve_default_training_metric_definitions(
...
 
-    instance_specific_metric_name: str
-    for instance_specific_metric_definition in instance_specific_metric_definitions:
-        instance_specific_metric_name = instance_specific_metric_definition["Name"]
-        default_metric_definitions = list(
-            filter(
-                lambda metric_definition: metric_definition["Name"]
-                != instance_specific_metric_name,
-                default_metric_definitions,
+    if instance_specific_metric_definitions:
```
Comment (Captainia) [inline src/sagemaker/jumpstart/artifacts/metric_definitions.py:99]: nit: 
```
if not instance_specific_metric_definitions:
  return default_metric_definitions
```

File: src/sagemaker/jumpstart/hub/parser_utils.py [hunk 4]
```diff
@@ -10,16 +10,20 @@
 # distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 # ANY KIND, either express or implied. See the License for the specific
 # language governing permissions and limitations under the License.
+# pylint: skip-file
```
Comment (Captainia) [inline src/sagemaker/jumpstart/hub/parser_utils.py:13]: nit: generally not a fan of skipping entire file, we should fix in a follow-up PR

File: src/sagemaker/jumpstart/hub/parser_utils.py [hunk 5]
```diff
@@ -29,20 +33,29 @@ def snake_to_upper_camel(snake_case_string: str) -> str:
...
             for key, value in json_obj.items():
                 new_key = apply(key)
-                if isinstance(value, dict):
-                    new[new_key] = {}
-                    _walk_and_apply_json(value, new=new[new_key])
-                elif isinstance(value, list):
-                    new[new_key] = []
-                    for item in value:
-                        _walk_and_apply_json(item, new=new[new_key])
+                if (stop_keys and new_key not in stop_keys) or stop_keys is None:
```
Comment (Captainia) [inline src/sagemaker/jumpstart/hub/parser_utils.py:49]: nit: 
```
if stop_keys and new_key in stop_keys:
    new[new_key] = value
else:
    ...
```

**Review by Captainia — APPROVED** [review]

File: src/sagemaker/jumpstart/hub/parser_utils.py (hunk 4 above)
Comment (malav-shastri) [reply src/sagemaker/jumpstart/hub/parser_utils.py:13]: me and @bencrabtree just discuss this, there's one apparently pylint false positive which we are not able to get rid of. Skipping it to unblock customer but sure I can followup

File: src/sagemaker/jumpstart/factory/model.py (hunk 1 above)
Comment (malav-shastri) [reply src/sagemaker/jumpstart/factory/model.py:262]: this is an optional field but looks like integ tests could be a possible solution here

Comment (bencrabtree) [reply src/sagemaker/jumpstart/factory/model.py:262]: Yes, we desperately need integ tests for (at least):

1. ModelReference fine-tuning
2. ModelReference deploy
3. Alt config deploy

File: src/sagemaker/jumpstart/hub/parser_utils.py [hunk 6]
```diff
@@ -10,16 +10,20 @@
...
 import re
-from typing import Any, Dict
+from typing import Any, Dict, List, Optional
 
 
 def camel_to_snake(camel_case_string: str) -> str:
     """Converts camelCaseString or UpperCamelCaseString to snake_case_string."""
     snake_case_string = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", camel_case_string)
+    if "-" in snake_case_string:
+        # remove any hyphen from the string for accurate conversion.
```
Comment (JGuinegagne) [inline src/sagemaker/jumpstart/hub/parser_utils.py:25]: nit: unnecessary comment

File: src/sagemaker/jumpstart/hub/parser_utils.py [hunk 7]
```diff
@@ -29,20 +33,29 @@ def snake_to_upper_camel(snake_case_string: str) -> str:
     return upper_camel_case_string
 
 
-def walk_and_apply_json(json_obj: Dict[Any, Any], apply) -> Dict[Any, Any]:
-    """Recursively walks a json object and applies a given function to the keys."""
+def walk_and_apply_json(
+    json_obj: Dict[Any, Any], apply, stop_keys: Optional[List[str]] = ["metrics"]
```
Comment (JGuinegagne) [inline src/sagemaker/jumpstart/hub/parser_utils.py:37]: non-blocking: apply typing please

File: src/sagemaker/jumpstart/hub/parser_utils.py (hunk 5 above)
Comment (JGuinegagne) [reply src/sagemaker/jumpstart/hub/parser_utils.py:49]: +1, conditional is too complex. If you want to keep this condition, create a utils and unit-test separately please.

**Review by pintaoz-aws — APPROVED** [review]

File: src/sagemaker/jumpstart/utils.py [hunk 8]
```diff
@@ -1103,6 +1104,17 @@ def get_jumpstart_configs(
...
+    if hub_arn:
+        return (
+            {
+                config_name: metadata_configs.configs[
+                    camel_to_snake(snake_to_upper_camel(config_name))
+                ]
+                for config_name in config_names
+            }
+            if metadata_configs
+            else {}
```
Comment (JGuinegagne) [inline src/sagemaker/jumpstart/utils.py:1116]: ternary + list-comprehension is a recipe for unreadability. Consider:
```py
if metadata_configs:
  return { ... }
else:
  return {}
```

**Pull request merged**